fix: ensure validation errors after async preValidation reach error handler (#6214)

emicovi opened the pull request:
When a custom validator throws inside the normal validation step _after_ an async preValidation hook, the error was bubbling out as an unhandled rejection instead of being forwarded to Fastify’s error handling.

This change wraps the internal `validateSchema()` invocation in a try/catch so any synchronous error is routed through `reply.send(err)` and handled by the configured `setErrorHandler`.

A regression test (`test/async-prevalidation-validation-error.test.js`) reproduces the scenario and asserts that a 400 response is returned via the custom error handler.

Closes #6214

emicovi committed:
fix: ensure validation errors after async preValidation reach error handler\n\nWhen a custom validator throws inside the normal validation step _after_ an async preValidation hook, the error was bubbling out as an unhandled rejection instead of being forwarded to Fastifys error handling.\n\nThis wraps the internal validateSchema invocation in a try/catch so any synchronous error is routed through reply.send(err).\n\nA regression test reproduces the scenario and asserts that a 400 response is returned via the custom setErrorHandler.\n\nCloses #6214

metcoder95 committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

Eomm commented on lib/handleRequest.js:
```diff
@@ -86,7 +86,16 @@ function preValidationCallback (err, request, reply) {
+    validationErr = validateSchema(reply[kRouteContext], request)
+  } catch (validationError) {
+    // Ensure synchronous validation errors are always routed through Fastify's error handling
+    reply[kReplyIsError] = true
+    reply.send(validationError)
+    return
```
we should use the `validationCompleted()` method to comply to the `attachValidation` option (and we need a test case for this as well)

emicovi replied:
@Eomm Fixed! Now using validationCompleted() instead of direct reply.send() and added a test case for the attachValidation option. All tests passing.

emicovi committed:
fix: use validationCompleted() to handle attachValidation option

- Replace direct reply.send(validationError) with validationCompleted() call
- Ensures attachValidation option works correctly with async preValidation hooks
- Add comprehensive test case for attachValidation scenario
- Maintains backward compatibility for existing error handling

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
Merge branch 'fix-6214-prevalidation-error-handler' of https://github.com/emicovi/fastify into fix-6214-prevalidation-error-handler

emicovi commented:
Yes I just did. Thank you!

emicovi committed:
fix: rename catch variable to avoid confusion with validationErr

- Rename validationError to thrownValidationErr in catch block
- Improves code clarity by avoiding similar variable names
- Addresses PR feedback from Uzlopak

emicovi commented:
Thanks for the feedback @Uzlopa I've renamed the catch variable from validationError to thrownValidationErr to avoid confusion with the existing validationErr variable.

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

climba03003 commented:
I would like to see if `Promise.try` helps in this case.
From my investigation, the higher the error rate. 
The performance between sync and async would be lower.
Ponyfill is the slowest in all case, around 10% lost.

```js
const Benchmark = require('benchmark')

function randomThrow() {
  if (Math.random() > 0.8) {
    throw Error()
  }
}

const suite = new Benchmark.Suite()
suite
  .add('sync try-catch', function () {
    try {
      randomThrow()
    } catch (err) { }
  })
  .add('Promise.try', {
    defer: true,
    fn(deferred) {
      Promise.try(randomThrow).then(() => deferred.resolve(), () => deferred.resolve())
    }
  })
  .add('Promise.try ponyfill', {
    defer: true,
    fn(deferred) {
      new Promise(function (resolve) {
        resolve(randomThrow())
      }).then(() => deferred.resolve(), () => deferred.resolve())
    }
  })
  .on("cycle", function (event) {
    console.log(String(event.target));
  })
  .on("complete", function () {
    console.log("Fastest is " + this.filter("fastest").map("name"));
  })
  .run({ async: true });
  ```

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

mcollina reviewed (approved):
lgtm

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

gurgunday reviewed (approved):
Can you add a TODO for Promise.try

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
Add TODO comment for Promise.try future enhancement

- Add TODO comment suggesting Promise.try when Node.js 20 support is dropped
- Addresses feedback from gurgunday in PR review
- Promise.try would provide cleaner async/sync error handling but requires Node.js 22.10+

gurgunday reviewed (approved):
lgtm

Eomm committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
test: add handling for synchronous errors in preValidation hook

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
refactor: simplify preValidation error handling in handler function

mcollina reviewed (approved):
lgtm

mcollina commented on lib/handleRequest.js:
```diff
@@ -86,7 +82,17 @@ function preValidationCallback (err, request, reply) {
-  const validationErr = validateSchema(reply[kRouteContext], request)
+  let validationErr
+  try {
+    // TODO: Consider using Promise.try when Node.js 20 support is dropped (requires Node.js 22.10+)
+    validationErr = validateSchema(reply[kRouteContext], request)
```
```suggestion
    validationErr = validateSchema(reply[kRouteContext], request)
```

@gurgunday note that `Promise.try` is a _bad idea_ here, as it would always add a microtick, which really we should not be doing if it's not needed (the vast majority of the case validation should not do any I/O).

mcollina reviewed (changes requested):
Sorry, we should be removing the added TODO, as it's really not a good idea to use it here.

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
refactor: remove outdated TODO comment regarding Promise.try in preValidationCallback

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

gurgunday reviewed (commented):
Can you resolve conflicts please?

Uzlopak committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

Eomm committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

Eomm committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

gurgunday reviewed (changes requested):
This regression test passes on `main`

I suppose it's because of https://github.com/fastify/fastify/pull/6247?

Is this PR still needed?

emicovi committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler

Eomm committed:
Merge branch 'main' into fix-6214-prevalidation-error-handler
